main/lsp/requests/code_action.cc

```diff
@@ -161,31 +162,61 @@ unique_ptr<ResponseMessage> CodeActionTask::runRequest(LSPTypecheckerDelegate &t
-            action->kind = CodeActionKind::RefactorExtract;
-
+        if (auto *def = hasLoneMethodResponse(gs, queryResult.responses)) {
+            unique_ptr<CodeAction> action;
             bool canResolveLazily = config.getClientConfig().clientCodeActionResolveEditSupport &&
                                     config.getClientConfig().clientCodeActionDataSupport;
-            auto newModuleLoc = getNewModuleLocation(gs, *def, typechecker);
-            auto renameCommand = make_unique<Command>("Rename Symbol", "sorbet.rename");
-            auto arg = make_unique<TextDocumentPositionParams>(
-                make_unique<TextDocumentIdentifier>(params->textDocument->uri), move(newModuleLoc));
-            auto args = vector<unique_ptr<TextDocumentPositionParams>>();
-            args.emplace_back(move(arg));
-
-            renameCommand->arguments = move(args);
-            action->command = move(renameCommand);
-            if (canResolveLazily) {
-                action->data = move(params);
+
+            if (def->symbol.data(gs)->owner.data(gs)->isSingletonClass(gs)) {
+                auto action = make_unique<CodeAction>("Move method to a new module");
+                action->kind = CodeActionKind::RefactorExtract;
+
+                auto newModuleLoc = getNewModuleLocation(gs, *def, typechecker);
+                auto renameCommand = make_unique<Command>("Rename Symbol", "sorbet.rename");
+                auto arg = make_unique<TextDocumentPositionParams>(
+                    make_unique<TextDocumentIdentifier>(params->textDocument->uri), move(newModuleLoc));
+                auto args = vector<unique_ptr<TextDocumentPositionParams>>();
+                args.emplace_back(move(arg));
+
+                renameCommand->arguments = move(args);
+                action->command = move(renameCommand);
+                if (canResolveLazily) {
+                    action->data = move(params);
+                } else {
+                    auto workspaceEdit = make_unique<WorkspaceEdit>();
+                    auto edits = getMoveMethodEdits(typechecker, config, *def);
+                    workspaceEdit->documentChanges = move(edits);
+                    action->edit = move(workspaceEdit);
+                }
+
+                result.emplace_back(move(action));
             } else {
-                auto workspaceEdit = make_unique<WorkspaceEdit>();
-                auto edits = getMoveMethodEdits(typechecker, config, *def);
-                workspaceEdit->documentChanges = move(edits);
-                action->edit = move(workspaceEdit);
+                auto action = make_unique<CodeAction>("Convert to singleton class method (best effort)");
+                action->kind = CodeActionKind::RefactorRewrite;
+
+                if (canResolveLazily) {
+                    const auto &maybeSource = def->termLoc.source(gs);
+                    if (maybeSource.has_value() && absl::StartsWith(maybeSource.value(), "def ")) {
```

Comment:
It has to be in both places, once here so that we don't even show the 💡 in VS Code (if we only had it in createMethodDefEdit, then it would not be called when lazily-resolving the code action, so VS Code would think there was a code action when there wasn't.

It might make sense to return an error from `createMethodDefEdit`, but for reasons mentioned elsewhere in comments in the PR, the behavior for an error from this code action would be to silently drop it anyways, so I'm not sure it's worth it.